src/Microsoft.AspNet.OData.Shared/Query/IODataQueryOptionsParser.cs

```diff
@@ -0,0 +1,33 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System.IO;
+#if !NETCORE
+using System.Net.Http.Formatting;
+#else
+using Microsoft.AspNet.OData.Formatter;
+#endif
+
+namespace Microsoft.AspNet.OData.Query
+{
+    /// <summary>
+    /// Exposes the ability to read and parse the content of a <see cref="T:System.IO.Stream" /> 
+    /// into a query options part of an OData URL. Query options may be passed 
+    /// in the request body to a resource path ending in /$query.
+    /// </summary>
+    public interface IODataQueryOptionsParser
+    {
+        /// <summary>
+        /// Gets the media type supported by the parser.
+        /// </summary>
+        MediaTypeMapping MediaTypeMapping { get; }
+
+        /// <summary>
+        /// Reads and parses the content of a <see cref="T:System.IO.Stream" /> 
+        /// into a query options part of an OData URL.
+        /// </summary>
+        /// <param name="requestStream">A <see cref="T:System.IO.Stream" /> containing the query options.</param>
+        /// <returns>A string representing the query options part of an OData URL.</returns>
+        string Parse(Stream requestStream);
+    }
+}
```

Comment:
In my opinion, instead of making the interface expose a single `MediaTypeMapping` that it supports, I think it should expose a method that encapsulates the logic of figuring out whether the parser can handle the request, e.g. `CanParse(context)` that returns a boolean. If the goal is to be extensible, I think this would be more flexible.

Furthermore, I think its consistent with the design of other AspNetCore concepts like input formatters:
```c#
interface IInputFormatter
{
   bool CanRead(InputFormatterContext);
   void ReadAsync(InputFormatterContext);
}
```

Reply:
After reading more of your code, I've realized that `MediaTypeMapping` does have a `TryMatchMediaType` which technically does what I was saying, so I don't know if it's still worth going with the `CanParse(context)` approach, but would like to hear your thoughts on this.

Reply:
@habbes Input and output formatters also use the `MediaTypeMapping` approach to determine how a resource should be materialized/serialized

Reply:
> In my opinion, instead of making the interface expose a single `MediaTypeMapping` that it supports, I think it should expose a method that encapsulates the logic of figuring out whether the parser can handle the request, e.g. `CanParse(context)` that returns a boolean. If the goal is to be extensible, I think this would be more flexible.
> 
> Furthermore, I think its consistent with the design of other AspNetCore concepts like input formatters:
> 
> ```cs
> interface IInputFormatter
> {
>    bool CanRead(InputFormatterContext);
>    void ReadAsync(InputFormatterContext);
> }
> ```

@habbes I implemented this suggestion